### PR TITLE
Feat: Safety Checks Setting

### PR DESCRIPTION
**Pull request by goodhoko**
Add a new feature that lets users to set the `safety_check` setting of their Trezor device.

### CODE REVIEW

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (szymonlesisz) [inline packages/suite/src/components/suite/Preloader/index.tsx]:
```suggestion
import { ModalRoute } from '@suite-constants/routes';
```
use aliases pls (and disable auto import plugin for this project)

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (szymonlesisz) [inline packages/suite/src/components/suite/Preloader/index.tsx]:
```suggestion
const getModalApplication = (route: ModalRoute) => {
```
unnecessary return type

File: packages/suite/src/components/suite/modals/SafetyChecks/index.tsx
Comment (szymonlesisz) [inline packages/suite/src/components/suite/modals/SafetyChecks/index.tsx]:
```suggestion
import { SafetyCheckLevel } from 'trezor-connect';
```

File: packages/suite/src/reducers/suite/deviceReducer.ts
Comment (szymonlesisz) [inline packages/suite/src/reducers/suite/deviceReducer.ts]: im thinking about the case where upcoming device is locked. 
upcoming features will not override current one therefore `features.unlocked` will have value `true` but it should have `false`

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (szymonlesisz) [reply packages/suite/src/components/suite/Preloader/index.tsx]: wouldn't be easier to have `router.route?.isModal` check inside this function and leave the rest as it was before?
introducing new type declared in different file and used only here looks kinda unnecessary

i mean something like previous line 98 `if (!route || !route.isModal) return;`

File: packages/suite/src/components/suite/modals/SafetyChecks/index.tsx
Comment (goodhoko) [reply packages/suite/src/components/suite/modals/SafetyChecks/index.tsx]: Thx. Fixed. This required [one additional change](https://github.com/trezor/connect/compare/d2bbc6a2207aed406be9082d8112a2c25603bfb2..75860f292720ed148a71ec8dde26e464eb3d6e3f) in the Connect PR to export `SafetyCheckLevel` at the root of Connect.

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (goodhoko) [reply packages/suite/src/components/suite/Preloader/index.tsx]: As explained in the [commit message ](https://github.com/trezor/trezor-suite/pull/3590/commits/517a147e0c6d1c393e023cdf06154be9bd4359cd) the motivation behind this is to make this function typesafe in a sense that TS compiler will barf if all possible cases aren't handled in the `switch`. This should help us in two ways:
1) Future devs won't spent time (as I did) figuring out why their newly added modal won't open with `goto()` when they didn't added a case to this switch.
2) It will prevent cases that can never happen and thus will keep the `switch` clean and minimal. This is exactly what happened with the `log` case. .)

For this to work we must tell the compiler that this function must always return a modal it will then ensure that the switch is exhaustive.

OTOH I agree that the return type *is* ugly. I'm open to any suggestion on making this nicer. `((...args: any[]) => JSX.Element | null)` is the only intersection of all our modals I could come up with. Maybe we can do better?

I also agree that definning `ModalRoute` could be done here. Either separately
```js
type ModalRoute = Extract<typeof routes[number], { isModal: true }>;
const getModalApplication = (route: ModalRoute): ...
```
or inline
```js
const getModalApplication = (route: Extract<typeof routes[number], { isModal: true }>): ...
```
I don't have a strong opinion on neither of these two.

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (goodhoko) [reply packages/suite/src/components/suite/Preloader/index.tsx]: Fixed. Resolving. Thx.

**Comment by goodhoko** [comment]
@szymonlesisz Thanks for an early review and some very good points.

I just wanted to stress that 1) this is work in progress and things will change 2) when I'll request for review I'll make it very explicit: I'll give tips how to review this: eg point specific people to specific commits, comment on parts I think deserve more attention or parts I'm not sure I did right.

File: packages/suite/src/reducers/suite/deviceReducer.ts
Comment (goodhoko) [reply packages/suite/src/reducers/suite/deviceReducer.ts]: Good point! 👍 With this PR the device could never become locked. What we probably need here is to only protect *some* of the features. We definitely do not want to protect `unlocked`. Are there any other features that actually *practically* change when the device gets locked?

Looking at the fields of `Features` interface, `session_id` seems (just a pure guess based on the name) like sommething statefull wrt being locked. 🤔

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (szymonlesisz) [reply packages/suite/src/components/suite/Preloader/index.tsx]: if you use suggested solution you will get exactly the same results without need to declare additional strict type.
[try it out here](https://www.typescriptlang.org/play?ssl=68&ssc=30&pln=67&pc=12#code/PTAEFp1BXAXBLANgZ1M2BDATrAULgUwA8AHAex1FgE8SDQBBLLDagUUQIFsCA7WADxMW1ACq16xWHwAmqLAQwyyvRNRi8A1rzIB3XgG0AugD5QAXkbNW4ugd7QuAIwJYjAbnzFylGnVAAwiroAOrwsAAWAPIkCCoYiABi8ASIcgK4oFmBwbCgUrKoAN6gBpoE1ABcaLBY8LwA5kbVGLzqAL4ANJnZyaly+UTSvAPodY2gAD6gDs6uU2jUzmSIuGaWRT1ZBgDSFaD1oOXUZABmObzozaCb2XelfWl76oePctdBl7C7FUYGb88jINhnItvc7tMxvUGmDwVlprMXFhYXDpiUyhVqlDGtdWh0UeDpk4yCtFLwCfcAPwXdA-ah-AG-Cl3arQEYEU71AgyTx3dqefn4EAQKBwJCoWRCsCQUBYMhwebobB4XAAY1ysvl0lQlgMYJKGBIJGqAHIdLBwFwyDIEuAAIwm0BdfWgQ3G0BmsgWq02xDgABMjuddwNRtN5st1ttAGYg90Q66wx6nHUZA0CCbOgdkABZKOIaq1aD0YPZUPuk3QGQEABumezed9hawxad8bLiYrNdcyHgKnr8Fz+ebredQIwqHVX08uD89AASlqCM8dUcKmcrCIONw+II5xu5QrkGZpibB42Eo7TyRsBguMgTZ5vBQ8nPQIuFRZN6xtzx+AJPlCcJoliPteASN5kAEfdzkPbUsw-aQVxME8NGrTleG5GdhRlOD5klWcJFAAAFOUSFXW5slVVpVVSDAnE4apiVJVpeWyU4SWqRFXDYrInGwLFamhAUpVlRRVTyQcZi9V0awwJB6M4UAIlcAgs04WBUBIBQQSoCIMFfCIpKkjBQHVLhyEw-g1Q1C9EDtL8AAptLIcjqlI1zkAASgsMxKKyHToCwXgPUSKIokdCczNydxQGFCg0DIHhQAAKQAZQADQAOl-XdcHafAp3QUA7P9JyXLckiyO83ybjBQLgo9AAhBh50iycYrisAEuQJL6HS7LcusgrCts-No3K6r3OqnzzD8+qCFgIKQocRBEFi+KsES5KBpyzg-zwEavFIF9oq+UB01gOyGCNRB4GouIQssRy8OqRCCFm+a7ngc5HIAQjwqZpgBpcsvPfMfIa3heLQXRwlVCJQBe0G3R8-z7mo5B6BNFN4DTDNKmZbIoZK-M7Rhu5MexqtaxNQm4XuEnSop7JhSpj1uywXt+3puFhTuJnxpnBnhR0UB0IwaBEDwPl8vcIA) and comment one of the switch cases.

turns out the real key to enable strict switch was removing default case :)

Regarding function return type, typescript will figure it out by itself and returned values will be strict (again, see example above)
returned components will require assigned props, like cancelable

i guess by resoving (2) we also resolve (1) as well :)

File: packages/suite/src/components/suite/modals/SafetyChecks/index.tsx
Comment (szymonlesisz) [reply packages/suite/src/components/suite/modals/SafetyChecks/index.tsx]: on second thought, you dont need to import or reexport anything.
typescript will resolve this type straight from device features so no additional action is required.

just use it like:
`const [level, setLevel] = useState(device.features?.safety_checks);`

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (goodhoko) [reply packages/suite/src/components/suite/Preloader/index.tsx]: > turns out the real key to enable strict switch was removing default case :)

Yes, but only when [`noImplicitReturns`](https://www.typescriptlang.org/tsconfig#noImplicitReturns) TS option is set to `true`. In Suite, we currently leave it to the default which is `false`. Enabling this lint raises some 65 errors in the Suite package. I'm up for doing that but in another PR. [Here](https://github.com/trezor/trezor-suite/issues/3630)'s an issue for that. Question is: do we want to keep the commit in question in this PR or defer it to the issue as well?

File: packages/suite/src/components/suite/modals/SafetyChecks/index.tsx
Comment (goodhoko) [reply packages/suite/src/components/suite/modals/SafetyChecks/index.tsx]: :+1: Edited (and also reverted the export of `SafetyCheckLevel`.

Pls check. .)

File: packages/suite/src/reducers/suite/deviceReducer.ts
Comment (goodhoko) [reply packages/suite/src/reducers/suite/deviceReducer.ts]: I [edited the commit](https://github.com/trezor/trezor-suite/compare/d1e8804fd220b5bee0a8f434fed04124d4d393b9..8ba10279313a82ae33ee360bd4c4dc70a27313b2) to preserve all features except `unlocked`. Pls check. I'm open to any suggestion on how to make that code cleaner.

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (szymonlesisz) [reply packages/suite/src/components/suite/Preloader/index.tsx]: i've just checked on your branch that `ModalRoute` type does not work for exactly same reason.
the only reason why it throws error here is because you've defined return type of the `getModalApplication` function.
problem here is this return type is not precise, it returns `JSX.Element` with `any` props, therefore `<ApplicationModal` (line 177) will accept `any` props (example: set `cancelable="foo"` - no error, but it should be)

File: packages/suite/src/components/suite/modals/SafetyChecks/index.tsx
Comment (szymonlesisz) [inline packages/suite/src/components/suite/modals/SafetyChecks/index.tsx]: shouldn't `PromptAlways` be represented by third radio button? 
looks like you are not able to set `PromptAlways` from here

File: packages/suite/src/reducers/suite/__fixtures__/deviceReducer.ts
Comment (szymonlesisz) [inline packages/suite/src/reducers/suite/__fixtures__/deviceReducer.ts]:
```suggestion
                    safety_checks: null,
```
tested case should reflect reality

**Comment by szymonlesisz** [comment]
and you probably want to resolve conflicts in connect patch

File: packages/suite/src/reducers/suite/deviceReducer.ts
Comment (szymonlesisz) [reply packages/suite/src/reducers/suite/deviceReducer.ts]: yes, this should work

File: packages/suite/src/components/suite/modals/SafetyChecks/index.tsx
Comment (goodhoko) [reply packages/suite/src/components/suite/modals/SafetyChecks/index.tsx]: Good eye! 👁️ This indeed intentional - the [linked issue](https://github.com/trezor/trezor-suite/issues/2242) mandates that `PromptAlways` is an advanced feature we don't want to expose to regular Suite users. It can still be set via command line using trezor-env.

[I added](https://github.com/trezor/trezor-suite/pull/3590/commits/c444dfda7a9178e75dde5fe712f6e20c3ed0b540) two comments that aim to clarify this in the code.

File: packages/suite/src/reducers/suite/__fixtures__/deviceReducer.ts
Comment (goodhoko) [reply packages/suite/src/reducers/suite/__fixtures__/deviceReducer.ts]: Edited, see [the fixup](https://github.com/trezor/trezor-suite/pull/3590/commits/814d5899fb53d06ca7e1537d13b3970c5e2e663e).

File: packages/suite/src/components/suite/Preloader/index.tsx
Comment (goodhoko) [reply packages/suite/src/components/suite/Preloader/index.tsx]: > the only reason why it throws error here is because you've defined return type of the getModalApplication function.

Yup, exactly.

> problem here is this return type is not precise

Aha! Well I could either edit the signature to
```typescript
const getModalApplication = (route: ModalRoute): ((a: InjectedModalApplicationProps) => JSX.Element | null) => {...}
```

or remove this commit from this PR and defer this refactor to later when `noImplicitReturns` is enabled and we don't have to hack around with explicit return types. I'm up for the later. What do you think?

Comment (szymonlesisz) [reply packages/suite/src/components/suite/Preloader/index.tsx]: noImplicitReturns PR would be the right place for that, i think

Comment (goodhoko) [reply packages/suite/src/components/suite/Preloader/index.tsx]: Ok, removed the commit. 

I force-pushed here, as it seemed over-complicated to have a revert fixup commit. Hope that's ok in this case. 👉 👈

**Comment by goodhoko** [comment]
> and you probably want to resolve conflicts in connect patch

We have to wait with this PR for relase of connect anyway. [This e2e test](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1178624589) won't succeed until that's done. (The reason behind that is described [over here](https://github.com/trezor/connect/pull/785#issuecomment-811789943)). Hence my plan was to wait for the release of Connect and *then* rebase getting rid of the patch file altogether.

File: packages/suite/src/reducers/suite/deviceReducer.ts
Comment (goodhoko) [inline packages/suite/src/reducers/suite/deviceReducer.ts]: I'm open to suggestion on how to write this nicer. The code feels cumbersome for me.

File: packages/suite/src/components/firmware/Button.tsx
Comment (goodhoko) [inline packages/suite/src/components/firmware/Button.tsx:34]: Is ConfirmButton on its own worth the component?

It feels wrong to share a button along with the text which is subject to localization: its usually much easier to translate texts locally.

File: packages/components/src/components/form/RadioButton/index.native.tsx
Comment (goodhoko) [inline packages/components/src/components/form/RadioButton/index.native.tsx]: @tomexx I have also some question reagrding this commit, but since you're the author here I'll direct them directly at you. .)

But to the point:

Why do we even bother with exporting the 'native' RadioButton when it's actually not implemented at all. Can't we rather leave this file out completely?

File: packages/components/src/components/form/RadioButton/index.tsx
Comment (goodhoko) [inline packages/components/src/components/form/RadioButton/index.tsx]: @tomexx Since we don't use the native[1] radio buttons here I was wondering how does this stand for accessibility. Or generally, what's our take on accessibility in Suite?

[1] `<input type="radio"...`

File: packages/components/src/components/typography/Warning/index.tsx
Comment (goodhoko) [inline packages/components/src/components/typography/Warning/index.tsx]: I'm not sure if placing this component in `typography` section makes sense.

File: packages/suite/src/support/messages.ts
Comment (goodhoko) [inline packages/suite/src/support/messages.ts]: Shouldn't the translations have better default text that is closer to what the actual values will be?

Or what's the best way to put there? I can imagine that an empty string as something more defnsive might be better in some cases.

**Review by goodhoko — COMMENTED** [review]
@keraf @tomexx @slowbackspace This is ready for review. I recommend going commit by commit and reading the commit messages. They provide some more detail.

I'd also appreciate you to take a look at the comments I left below.

I'd also like to ask @tomexx and/or @slowbackspace to review the jsx parts of https://github.com/trezor/trezor-suite/pull/3590/commits/67bb2f93608f61bef697403112b6a5c712855575 and https://github.com/trezor/trezor-suite/pull/3590/commits/78a9f4601ee1250a28907bdfe3180ec89aba2c88 I don't feel strong about my html/css constructs.

I tagged you 'cos you were all invovled at least a bit by helping me directly or sharing advice. However there's no need for this PR to be examined by four people. This was also already seen by @szymonlesisz.  So don't spend a day on this. 🙏

File: packages/components/src/components/form/RadioButton/index.native.tsx
Comment (tomexx) [reply packages/components/src/components/form/RadioButton/index.native.tsx]: It's a placeholder for the native build. But I guess we can remove it as we don't have native app in development.

File: packages/components/src/components/typography/Warning/index.tsx
Comment (tomexx) [reply packages/components/src/components/typography/Warning/index.tsx]: I think we could put it as a standalone component. I'd also didn't expect it to be placed in typography.

File: packages/suite/src/support/messages.ts
Comment (goodhoko) [reply packages/suite/src/support/messages.ts]: Answered on Slack.

> Shouldn't the translations have better default text that is closer to what the actual values will be?

They should. I'm already putting together good defaults for the keys in this PR with other colleagues.

> Or what's the best way to put there? I can imagine that an empty string as something more defnsive might be better in some cases.

For the time being the attitude is that the dev writes a text that is good enough and doesn't bother with polishing the copy. That is left for someone who can do it properly later via Crowdin.

File: packages/suite/src/components/firmware/Button.tsx
Comment (tomexx) [reply packages/suite/src/components/firmware/Button.tsx:34]: I think it's OK to place it here to avoid inconsistency in the PR.

File: packages/suite/src/support/messages.ts
Comment (tomexx) [reply packages/suite/src/support/messages.ts]: I think it there has to be some meaningful content but also might be changed later from Crowding and be synced back to the app. I personally wouldn't use an empty string.

Comment (goodhoko) [reply packages/suite/src/support/messages.ts]: A good enough translations added in https://github.com/trezor/trezor-suite/pull/3590/commits/91681f244202c2730d433b7aef8d89466d999c4c.

Resolving.

File: packages/suite/src/components/firmware/Button.tsx
Comment (goodhoko) [reply packages/suite/src/components/firmware/Button.tsx:34]: Ok fine. .) Resolving.

File: packages/components/src/components/form/RadioButton/index.native.tsx
Comment (goodhoko) [reply packages/components/src/components/form/RadioButton/index.native.tsx]: Ok. removed in https://github.com/trezor/trezor-suite/pull/3590/commits/023a1adfd80e4a4bcbb0822e8b990e5d1ca9f234. Resolving.

File: packages/suite/src/reducers/suite/deviceReducer.ts
Comment (goodhoko) [reply packages/suite/src/reducers/suite/deviceReducer.ts]: No suggestions.

_sigh_

Resolving.

File: packages/components/src/components/typography/Warning/index.tsx
Comment (goodhoko) [reply packages/components/src/components/typography/Warning/index.tsx]: @tomexx where would you put it? .)

File: packages/components/src/components/form/RadioButton/index.tsx
Comment (goodhoko) [reply packages/components/src/components/form/RadioButton/index.tsx]: Sigh. Moving forward and resolving.

Comment (tomexx) [reply packages/components/src/components/form/RadioButton/index.tsx]: So far we did not prioritized usability. So I'd keep it like this for now.

File: packages/components/src/components/typography/Warning/index.tsx
Comment (tomexx) [reply packages/components/src/components/typography/Warning/index.tsx]: Let's put it as a standalone component `packages/components/src/components/Warning` we can refactor later when new elements/variations will be introduced.

Comment (goodhoko) [reply packages/components/src/components/typography/Warning/index.tsx]: Alright. Moved in https://github.com/trezor/trezor-suite/pull/3590/commits/283b5e07c27399bcbf2dc75a637a72de62a36922. Thanks!

Resolving.

**Comment by goodhoko** [comment]
The failing tests doesn't seem related to this PR. Can we re-run them?

Also this is ready to be merged from my side, so if anyone with the priveleges could hit that button... 👉 👈

**Review by szymonlesisz — APPROVED** [review]

**Pull request merged**
